Shoko.Server/Models/Release/StoredReleaseInfo.cs

```diff
@@ -0,0 +1,285 @@
+
+    public DateTime CreatedAt { get; set; }
+
+    long? IReleaseInfo.FileSize => FileSize;
+
+    public IReadOnlyList<IReleaseVideoCrossReference> CrossReferences
+    {
+        get => EmbeddedCrossReferences
+            .Split(',')
+            .Select(EmbeddedCrossReference.FromString)
+            .WhereNotNull()
+            .ToList();
+        set => EmbeddedCrossReferences = value
+            .Select(x => x.ToEmbeddedString())
+            .Join(',');
+    }
+
+    #endregion
+
+    #region IReleaseGroup Implementation
+
+    public string? GroupID { get; set; }
+
+    public string? GroupSource { get; set; }
+
+    public string? GroupName { get; set; }
+
+    public string? GroupShortName { get; set; }
+
+    public bool Equals(IReleaseGroup? other)
+        => other is not null &&
+            string.Equals(GroupID, other.ID) &&
+            string.Equals(GroupSource, other.Source);
+
+    IReleaseGroup? IReleaseInfo.Group => !string.IsNullOrEmpty(GroupID) && !string.IsNullOrEmpty(GroupSource) && !string.IsNullOrEmpty(GroupName) && !string.IsNullOrEmpty(GroupShortName) ? this : null;
+
+    string IReleaseGroup.ID => GroupID ?? string.Empty;
+
+    string IReleaseGroup.Source => GroupSource ?? string.Empty;
+
+    string IReleaseGroup.Name => GroupName ?? string.Empty;
+
+    string IReleaseGroup.ShortName => GroupShortName ?? string.Empty;
+
+    #endregion
+
+    #region IReleaseMediaInfo Implementation
+
+    public string? EmbeddedAudioLanguages { get; set; }
+
+    public string? EmbeddedSubtitleLanguages { get; set; }
+
+    public IReadOnlyList<TitleLanguage>? AudioLanguages
+    {
+        get => EmbeddedAudioLanguages?.Split(',').Select(l => l.Trim().GetTitleLanguage()).ToList();
+        set => EmbeddedAudioLanguages = value?.Select(l => l.GetString()).Join(",");
+    }
+
+    public IReadOnlyList<TitleLanguage>? SubtitleLanguages
+    {
+        get => EmbeddedSubtitleLanguages?.Split(',').Select(l => l.Trim().GetTitleLanguage()).ToList();
+        set => EmbeddedSubtitleLanguages = value?.Select(l => l.GetString()).Join(",");
+    }
+
+    IReleaseMediaInfo? IReleaseInfo.MediaInfo => AudioLanguages is not null || SubtitleLanguages is not null ? this : null;
+
+    IReadOnlyList<TitleLanguage> IReleaseMediaInfo.AudioLanguages => AudioLanguages ?? [];
+
+    IReadOnlyList<TitleLanguage> IReleaseMediaInfo.SubtitleLanguages => SubtitleLanguages ?? [];
+
+    #endregion
+
+    #region IHashes Implementation
+
+    public string? Hashes { get; set; }
+
+    string IHashes.ED2K => this[HashAlgorithmName.ED2K] ?? string.Empty;
+
+    string? IHashes.MD5 => this[HashAlgorithmName.MD5];
+
+    string? IHashes.SHA1 => this[HashAlgorithmName.SHA1];
+
+    string? IHashes.CRC => this[HashAlgorithmName.CRC32];
+
+    public string? this[HashAlgorithmName algorithm]
+    {
+        get
+        {
+            var (ed2k, md5, sha1, crc32) = Hashes?.Split('|', StringSplitOptions.TrimEntries) ?? [];
```

Comment:
I'm not sure this will work. We can't guarantee that a plugin provided hash won't contain a pipe. I would store these in a separate table with a string Type, string or blob Hash, and any IDs necessary to link

Reply:
Maybe a few columns for "extra data". Kind of like how some hashes require a salt or whatever to reproduce.

Reply:
Additionally why don't we provide a FileHash table which is a combination of:
```
FileID   | int FK | PK
HashType | string | PK
Hash     | String |
```

Reply:
Presumably, the point of abstracting the hashes is so that new hash types could be added, such as perceptual hashes. In those cases, the extra data would be timestamps. 
The extra data would be on the hash object. The hash object would replace the hashes everywhere a hash is relevant

Reply:
so if I'm understanding it right, then we add a new hash object (hash type, hash value, extra data) and convert all occurrences of the existing `IHashes` interface to instead expose a list of these new hash objects?